lib/RateLimiterDynamo.js

```diff
@@ -0,0 +1,422 @@
+        Key: {
+          key: {S: rlKey}
+        },
+        ConditionExpression: 'attribute_exists(#k)',
+        ExpressionAttributeNames: {
+          '#k': 'key'  
+        }
+      }
+      
+      try {
+        const data = await this._client.deleteItem(params);
+        return data.$metadata.httpStatusCode === 200;
+      } catch(err) {
+        // ConditionalCheckFailed, item does not exist in table
+        if (err.__type.includes('ConditionalCheckFailedException')) {
+          return false;
+        } else {
+          throw err;
+        }
+      }
+
+    }
+
+    /**
+     * Implemented with DynamoDB Atomic Counters. 3 calls are made to DynamoDB but each call is atomic.
+     * From the documentation: "UpdateItem calls are naturally serialized within DynamoDB,
+     * so there are no race condition concerns with making multiple simultaneous calls."
+     * See: https://aws.amazon.com/it/blogs/database/implement-resource-counters-with-amazon-dynamodb/
+     * @param {*} rlKey 
+     * @param {*} points 
+     * @param {*} msDuration 
+     * @param {*} forceExpire 
+     * @param {*} options 
+     * @returns
+     */
+    async _upsert(rlKey, points, msDuration, forceExpire = false, options = {}) {
+
+      if (!this.tableCreated) {
+        throw new Error('Table is not created yet');
+      }
+
+      const dateNow = Date.now();
+      /* -1 means never expire, DynamoDb do not support null values in number fields.
+         DynamoDb TTL use unix timestamp in seconds.
+      */
+      const newExpireSec = msDuration > 0 ? (dateNow + msDuration) / 1000 : -1;
+
+      // Force expire, overwrite points. Create a new entry if not exists
+      if (forceExpire) {
+        return await this._baseUpsert({
+          TableName: this.tableName,
+          Key: { key: {S: rlKey} },
+          UpdateExpression: 'SET points = :points, expire = :expire',
+          ExpressionAttributeValues: {
+            ':points': {N: points.toString()},
+            ':expire': {N: newExpireSec.toString()}
+          },
+          ReturnValues: 'ALL_NEW'
+        });
+      }
+
+      // First try update, success if entry does not exists
+      try {
+        
+        return await this._baseUpsert({
+          TableName: this.tableName,
+          Key: { key: {S: rlKey} },
+          UpdateExpression: 'SET points = :new_points, expire = :new_expire',
+          ExpressionAttributeValues: {
+            ':new_points': {N: points.toString()},
+            ':new_expire': {N: newExpireSec.toString()},
+          },
+          ConditionExpression: 'attribute_not_exists(points)',
+          ReturnValues: 'ALL_NEW'
+        });
+
+      } catch (err) {
+
+        // Second try update, success if entry exists and is not expired
+        try {
+
+          return await this._baseUpsert({
+            TableName: this.tableName,
+            Key: { key: {S: rlKey} },
+            UpdateExpression: 'SET points = :new_points, expire = :new_expire',
```

Comment:
@Daniel-97 Should it be `SET points = points + :new_points`? This update should summarize points with existing points in DynamoDB. The total number of points in the time window will be the number to rely on when limiting points consumption.

Reply:
As i mention in my last comment, the second update means the key is expired. Let me know what you think